src/backend/distributed/executor/multi_utility.c

```diff
@@ -1016,13 +1083,199 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
 				break;
 			}
 
+			case AT_AddConstraint:
+			{
+				Constraint *constraint = (Constraint *) command->def;
+				LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
+				Oid referencingTableId = InvalidOid;
+				Oid referencedTableId = InvalidOid;
+				Var *referencingTablePartitionColumn = NULL;
+				Var *referencedTablePartitionColumn = NULL;
+				ListCell *referencingTableAttr = NULL;
+				ListCell *referencedTableAttr = NULL;
+				bool foreignConstraintOnPartitionColumn = false;
+
+				/*
+				 * We do not allow creation of a constraint if we are in a middle of a
+				 * transaction. Because other queries in the transaction may connect to
+				 * the workers from different connection and obtain conflicting locks
+				 * with this query.
+				 */
+				if (XactModificationLevel > XACT_MODIFICATION_NONE)
```

Comment:
It doesn't seem like we still need this check. Multi-shard commands now always use the same connection for all co-located placements, so it should be safe to do other multi-shard commands in the same transaction as adding a foreign key. If there is a preceding single-shard command then `ExecuteDistributedForeignKeyCommand `/`ExecuteModifyTasks` will error out.